Added skin descriptor fetching by UUID

Mqzn opened the pull request:
# Implementation changes:-
Refactored `FetchingDescriptor` into `NameFetchingDescriptor`
Added `UUIDFetchingDescriptor` for fetching skins from uuid.

# API changes
Added API method `SkinDescriptorFactory#createRefreshingDescriptor(UUID playerUUID)`

Mqzn committed:
Added skin descriptor fetching by UUID

Pyrbu commented:
You need to change the `serialize()` method of your new descriptor and add a matching deserializer in the `BaseSkinDescriptor.deserialize()` method, the way your code currently is it will treat the UUID as a player name on server restart.

Mqzn committed:
Adapted the serialization process to the new fetching descriptors

Mqzn commented:
> You need to change the `serialize()` method of your new descriptor and add a matching deserializer in the `BaseSkinDescriptor.deserialize()` method, the way your code currently is it will treat the UUID as a player name on server restart.

Done !
- Changed the serialization string to the format **"fetching-name"** for name fetcher and **"fetching-uuid"** for uuid fetcher.
- Added a check in `BaseSkinDescriptor.deserialize()` to return an instance of uuid/name fetcher depending on the serialization string.

Pyrbu commented:
> Changed the serialization string to the format "fetching-name" for name fetcher

The issue with doing that is that old NPC files will fail to load, so you need to change that one back or add a second check in the deserialization method that checks for the old format.

Mqzn committed:
Made it compatible with old npc files

Mqzn commented:
Done; Made it compatible by checking if the associated serialization string component is a uuid or not
if the string is a uuid -> return an instance of the uuid fetcher, otherwise it's a name.

Mqzn commented:
Is there anything left to do ?

Pyrbu commented:
I would prefer it if you just changed back the old fetching descriptor to fetching and called the UUID one something like `fetching-uuid` since what you did here adds a bunch of unnecessary code, even if small adds overhead and makes the YAML file harder to understand since two different types have the same "id"

Mqzn committed:
Removed unnecessary code

Mqzn commented:
> I would prefer it if you just changed back the old fetching descriptor to fetching and called the UUID one something like `fetching-uuid` since what you did here adds a bunch of unnecessary code, even if small adds overhead and makes the YAML file harder to understand since two different types have the same "id"

Alright, I changed it back and called the uuid one `fetching-uuid`, and i added a check for it in the `BaseSkinDescriptor#deserialize` method

Pyrbu commented:
Alright, LGTM!
